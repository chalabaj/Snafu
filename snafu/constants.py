import numpy as np
# CONVERSION CONSTANTS
AU_FS = 0.02418884326505e0   # atomic units to femtosecs
AU_EV = 27.21138386
AMU = 1822.888484264545e0    # atomic mass unit me = 1 AMU*atomic weight
ANG_BOHR = 1.889726132873e0  # agstroms to bohrs
HBAR_EV = 6.58211889e-16
HBAR_AU = 1.00000e0
BOHR_ANG = 1/ANG_BOHR
np.set_printoptions(precision=5)  # for print and stability testing
