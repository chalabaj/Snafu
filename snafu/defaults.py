#  DEFAULT VALUES 
#  might be overwriten by input.in values
ener_thresh = 1.000
hop_thresh = 0.5
dt = 4.00
tera_mpi = 0
write_freq = 10
restart_freq = 100
vel_adj = 0
timestep = 4
method = "lz"
hop = False
step = 0
dE = 0.0  # energy change since initial energies
Etot_init = 0.0  # setting variable , total energy at the beginning
Etot_prev = 0.0
sim_time = 0.0
prob = 0.0
