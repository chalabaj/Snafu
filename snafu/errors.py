
import sys
import os

def error_exit(error_number, error_desc=" "):
    """
    Exiting SNAFU simulation.
    MPI4PY implementation of MPI is currently not perfect when dealing with errors and can cause deadlocks. Hence MPI checking.
    """
    #  tera_mpi = 0
    err = ("0 - File input.in not found in folder.\n",
           "1 - File geom.in not found in folder.\n",
           "2 - Number of atoms in geoms.in and input.in is not consistent. Please check geom.in and input.in and check XYZ format.\n",
           "3 - OS error: {0}".format("Could not create necessarry files.\n"),
           "4 - Ab initio calculations failed.\n",
           "5 - Ab initio interface (r.X X=gauss, molpro etc.) not found. Check ABINITIO folder.\n",
           "6 - Hopping probability larger than 1, something went wrong.\n",
           "7 - Too large energy drift.\n",
           "8 - File {} exists, but the restart option is turned off (restart = 0). Either remove file or change the restart option.\n".format(error_desc),
           "9 - Input varible(s) is not properly set. See input.in.\n{}".format(error_desc),
           "10 - Restart file {} was not found.".format(error_desc),
           "11 - Wrong input parameter.\nError:{}.".format(error_desc),
           "12 - Input variable {} not set.".format(error_desc),
           "13 - Error in TeraChem interface.Error:{}".format(error_desc),
           "14 - Wrong format of geom.in or veloc.in. Should be XYZ format with geometry in Angstrom and velocities in atomic units.\nError:{}.".format(error_desc),
           "15 - {}".format(error_desc),
           "16 - Error when reading data from restart file.",
           "17 - Error when reading restart file. Option \"{}\" not found.".format(error_desc),
           "18 - User called ABORT {}.".format(error_desc),
           "19 - Import error: {}.".format(error_desc)
          )
    print("-------------------------------------------------------------------")
    print(err[error_number])
    print("Program was terminated due to an error!")
    sys.stdout.flush() 
    
    # prevent MPI deadlock without raiseing runtime error which is caught by excepthook, does not create traceback
    try:
        tera_mpi = int(os.environ['MPI_TERA'])
    except Exception:
        tera_mpi = 0
    if tera_mpi:
         from tera_propagates import global_except_hook
         sys.excepthook = global_except_hook 
         raise RuntimeError("USER/INPUT ERROR.")
    sys.exit(1)    
    return()
