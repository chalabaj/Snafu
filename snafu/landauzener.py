
try:
    import numpy as np  
    import random  
except ImportError as ime:
    if ime.name is None:  # module could have been removed or module file renamed
        print("Import in some of the modules ({}) in snafu dir failed. Exiting...".format(ime.name))
        exit(1)

      
def calc_lz_hopp(method, state, pot_eners,
                 pot_eners_array, Ekin, dt):
    
                
    hop = False             
    instate = state
    # add last calculated pot energies to evaluate hop 
    # based on three point minima
    pot_eners_array = np.append(pot_eners_array, pot_eners, axis = 0)
    print(pot_eners_array)
    
    prob = [calc_prob(instate, outstate, pot_eners_array) for outstate in range(0,len(pot_eners) if outstate <> instate ]
    print(minima)
    
    
    max_prob = (np.argmax(prob, axis = 0)[3]) # find row with max probaBILITY 
    outstate = prob[max_prob][1]
    
    if prob[max_prob][3] > random.random()
      dEpot = pot_eners_array[2][outstate] - pot_eners_array[2][instate]
        if prob[max_prob][2] < Ekin  
            hop = True
     
    if conservation
    
    if more than 1 possibilities
    highest prob wins 
      compare with random number 
      
    
    
    # hop to another PES, remove appended energies
    if hop:
        pot_eners_array = np.delete(pot_eners_array, 0, axis = 0)  
    else:
        pot_eners_array = np.delete(pot_eners_array, -1, axis = 0)
    
    return(hop, pot_eners_array)
   
def calc_prob(instate, outstate, pot_eners_array, dt):

    hbar_ev = 6.58211889e-16
    minima = False

    Z = [ (abs(pot_eners_array[step][instate] - pot_eners_array[step][oustate]) * au_eV) for step in range(1,4)]  # 1: T-DT, 2: T, 3: T+DT
    
    print(Z)
    if Z[0] > Z[1] and Z[1] < Z[2]:
        minima = True      
        print("Calculating hop prob. between {} - {} state with dE = {} ".format(minima[0], minima[1], minima[2]))
        sec_der = ((Z[2] - 2 * Z[1] + Z[0]) / (dt ** 2))
        prob = math.exp(-math.pi/(2 * hbar_ev)*(math.sqrt( Z[1]**3 / sec_der ) ))
        print(prob)
    return([instate, outstat, Z[1], prob])
