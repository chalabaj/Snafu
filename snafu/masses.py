# -*- coding: utf-8 -*-
"""
Created on Tue Jul 10 11:29:42 2018

@author: chalabaj
"""
import numpy as np
from constants import AMU
    
def assign_masses(at_names):
    """
    Atomic weights of the elements 2013 (IUPAC Technical Report) Meija, Juris; et al.
    http://doi.org/10.1515/pac-2015-0305, Pure and Applied Chemistry, 88, 3, pp. 265-291, 2016-02-24
    """
    
    atomic_masses = {    'H'    : 1.00794,	
                         'D'		:	2.01410177785,                   	
                         'He'		:	4.002602	,
                         'Li'		:	6.93800,	
                         'Be'		:	9.0121831	,
                         'B'		:	10.811	,
                         'C'		:	12.011	,
                         'N'		:	14.00674	,
                         'O'		:	15.999	,
                        'F'	  	:	18.9984032	,
                        'Ne'		:	20.1797	,
                        'Na'		:	22.989768	,
                        'Mg'		:	24.305	,
                        'Al'		:	26.981539	,
                        'Si'		:	28.0855	,
                        'P'		  :	30.973762	,
                        'S'		  :	32.066	,
                        'Cl'		:	35.4527	,
                        'Ar'		:	39.948	,
                        'K'	  	:	39.0983	,
                        'Ca'		:	40.078	,
                        'Sc'		:	44.95591	,
                        'Ti'		:	47.88	,
                        'V'	  	:	50.9415	,
                        'Cr'		:	51.9961	,
                        'Mn'		:	54.93805	,
                        'Fe'		:	55.847	,
                        'Co'		:	58.9332	,
                        'Ni'		:	58.6934	,
                        'Cu'		:	63.546	,
                        'Zn'		:	65.39	,
                        'Ga'		:	69.723	,
                        'Ge'		:	72.61	,
                        'As'		:	74.92159	,
                        'Se'		:	78.96	,
                        'Br'		:	79.904	,
                        'Kr'		:	83.8	,
                        'Rb'		:	85.4678	,
                        'Sr'		:	87.62	,
                        'Y'		  :	88.90585	,
                        'Zr'		:	91.224	,
                        'Nb'		:	92.90638	,
                        'Mo'		:	95.94	,
                        'Tc'		:	-97.9072	,
                        'Ru'		:	101.07	,
                        'Rh'		:	102.9055	,
                        'Pd'    :	106.42	,
                        'Ag'    :	107.8682	,
                        'Cd'	  :	112.411	,
                        'In'   	:	114.818	,
                        'Sn'   	:	118.71	,
                        'Sb'	  :	121.757	,
                        'Te'		:	127.6	,
                        'I'     :	126.90447	,
                        'Xe'		:	131.29	,
                        'Cs'		:	132.90543	,
                        'Ba'		:	137.327	}

    masses = np.zeros(len(at_names),dtype=np.float64)  
    
    for iat in range(0,len(at_names)):
       masses[iat] = (atomic_masses[at_names[iat]]) 
    am = masses*AMU   # atomic mass units conversion
    return(am)