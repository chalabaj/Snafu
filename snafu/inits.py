"""
Check input file: input.in, geom.in (input geometry), veloc.in pab initio program file,
Read input.in

each array (x,y,z) or (vx,vy,vz) or (fx,fy,fz)
holds NATOMS elements 
e.g. x[0] is x position of 1st atom, fx[1] force in x direction on 2 atom, etc.
"""

import sys
import os
import configparser
import numpy as np

try:
    from errors import error_exit
except ImportError as ime:
    print("Module {} not found.".format(ime.name))
    exit(1)

au_fs = 0.02418884326505e0      #atomic units to femtosecs
au_eV = 27.21139
amu   = 1822.888484264545e0             # atomic mass unit  me = 1 AMU*atomic weight
ang_bohr = 1.889726132873e0     # agstroms to bohrs

def file_check(cwd):
    # input files names - these are defaults otherwise not found
    veloc_file = "veloc.in"
    geom_file  = "geom.in"
    input_file = "input.in"
    
    # absolute path to the files  
    input_file_path = os.path.join(cwd,input_file)
    geom_file_path  = os.path.join(cwd,geom_file)
    vel_file_path = os.path.join(cwd,veloc_file)
    if (not os.path.isfile(input_file_path)):
         error_exit(0)
    if (not os.path.isfile(geom_file_path)):
         error_exit(1)
    if (not os.path.isfile(vel_file_path)):
        print("No initial velocities.")
        init_vel = 0
    else:
        print("Initial velocities read from {}".format(veloc_file)) 
        init_vel = 1
    
    return(input_file_path, geom_file_path, vel_file_path, init_vel)
    
def read_input(cwd,input_file_path):
    
    # Read parameter from input.in file
    cfg = configparser.ConfigParser(delimiters=('=', ':'),
        comment_prefixes=('#', ';'))
    cfg.read(input_file_path)       # Read file
    par=dict(cfg.items("Settings",raw=False))
    
    # To get rid of inline comments [0] and strip spaces
    print("Simulation will start with parameters: ")
    print("\n".join("{}: {}".format(k.strip(), v.split("#")[0].strip()) for k, v in par.items()))
    for p in par:
       par[p]=par[p].split("#",1)[0].strip(" ")  
        
    # Now that we know everything - check for ab initio interface
    abinit_file = "ABINITIO/{}.sh".format(par['abinitio'])
    ab_initio_file_path  = os.path.join(cwd,abinit_file)
    
    if (not os.path.isfile(ab_initio_file_path)):
         error_exit(5)
         
    return(par,ab_initio_file_path)

def read_geoms(natoms,geom_file_path):
    #if restart == 1 : read  last two geoms
    # could be xyz matrix [natoms,3] as well, but this should be more clear reading
    x = np.zeros(natoms, dtype=np.float64)  
    y = np.zeros(natoms, dtype=np.float64)  
    z = np.zeros(natoms, dtype=np.float64)  
    
    x_new = np.zeros(natoms, dtype=np.float64)  
    y_new = np.zeros(natoms, dtype=np.float64)  
    z_new = np.zeros(natoms, dtype=np.float64)  
    at_names = []
 # READ INITIAL POSITIONS:   
    with open(geom_file_path,'r') as igf:  # igf input geom file 
     atoms = igf.readline()  # first line in geom file is number of atoms
     if not (int(atoms) == natoms):  error_exit(2)                         
     garbage = igf.readline()  # comment 
     for iat in range(0,natoms):
        line = igf.readline().split()
        at_names.append(capitalize_2th(str(line[0])))
        x[iat] = np.float64(line[1]) * ang_bohr  # atomic units : Bohr
        y[iat] = np.float64(line[2]) * ang_bohr
        z[iat] = np.float64(line[3]) * ang_bohr
     
     igf.close()
     return(at_names, x, y, z, x_new, y_new, z_new)
 # READ INITIAL VELOCITIES:
def read_velocs(init_vel, natoms, vel_file_path):  
    vx = np.zeros(natoms, dtype=np.float64)  
    vy = np.zeros(natoms, dtype=np.float64)  
    vz = np.zeros(natoms, dtype=np.float64)  
    if init_vel:
        with open(vel_file_path,'r') as ivf:  
            atoms = ivf.readline()  # first line is number of atoms
            if not (int(atoms) == natoms):  
                error_exit(2)                         
            garbage = ivf.readline()  # second comment line
            for iat in range(0,natoms):
                line = ivf.readline().split()
                vx[iat] = float(line[1])
                vy[iat] = float(line[2])
                vz[iat] = float(line[3])
        ivf.close()
    return(vx, vy, vz)

    
def capitalize_2th(s):
    # Capitalizace first letter, lower second -
    # avoid problems with different name for atoms in ab initio codes
    return s[:1].capitalize() + s[1:].lower()

def init_forces(natoms,nstates):
    # Initialize empty forces array
    
    #f(t)
    fx = np.zeros(natoms, dtype=np.float64)  
    fy = np.zeros(natoms, dtype=np.float64)   
    fz = np.zeros(natoms, dtype=np.float64)  
    
    #f_new(t+dt)
    fx_new = np.zeros(natoms, dtype=np.float64)    
    fy_new = np.zeros(natoms, dtype=np.float64)  
    fz_new = np.zeros(natoms, dtype=np.float64)  
        
    return(fx, fy, fz, fx_new, fy_new, fz_new)
    
def init_energies(nstates):
    # Initialize empty potential energies array
    pot_eners = np.zeros(nstates, dtype=np.float64)     
    return(pot_eners)
    
def com_removal(x,y,z,am):
    totmass, xsum, ysum, zsum = 0.0, 0.0, 0.0, 0.0
    for iat in range(0,len(x)):
        xsum += x[iat] * am[iat] 
        ysum += y[iat] * am[iat] 
        zsum += z[iat] * am[iat] 
        totmass += am[iat] 
        
    xcom = xsum / totmass
    ycom = ysum / totmass
    zcom = zsum / totmass
     
    for iat in range(0,len(x)):
        x[iat] = x[iat] - xcom
        y[iat] = y[iat] - ycom
        z[iat] = z[iat] - zcom 
    
    return(x,y,z)

