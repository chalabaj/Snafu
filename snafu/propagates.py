"""
Velocity verlet integrator scheme
"""

import math
import sys
import os
import random
import time
import subprocess
import re
import numpy as np

try:
    from errors import error_exit
    from prints import print_energies, print_pes
    from constants import *
except ImportError as ime:
    print("Module {} not found.".format(ime.name))
    exit(1)
        
# CONSTANTS
au_fs = 0.02418884326505e0   # atomic units to femtosecs
au_eV = 27.21139
amu = 1822.888484264545e0    # atomic mass unit me = 1 AMU*atomic weight
ang_bohr = 1.889726132873e0  # agstroms to bohrs

def update_positions(dt, am, x, y, z, vx, vy, vz, fx, fy, fz):
    for iat in range(0,len(am)): 
        x[iat] = x[iat] + vx[iat] * dt + ( fx[iat] / (2 * am[iat]) * dt ** 2 )
        y[iat] = y[iat] + vy[iat] * dt + ( fy[iat] / (2 * am[iat]) * dt ** 2 )
        z[iat] = z[iat] + vz[iat] * dt + ( fz[iat] / (2 * am[iat]) * dt ** 2 )
    return(x,y,z)

def update_velocities(dt,am,vx,vy,vz,fx,fy,fz,fx_new,fy_new,fz_new):
    for iat in range(0,len(am)):
        vx[iat] = vx[iat] + ( dt * (fx[iat] + fx_new[iat]) / (2 * am[iat]) )
        vy[iat] = vy[iat] + ( dt * (fy[iat] + fy_new[iat]) / (2 * am[iat]) )
        vz[iat] = vz[iat] + ( dt * (fz[iat] + fz_new[iat]) / (2 * am[iat]) )
     #print(vz[iat])
    return(vx,vy,vz)   

def adjust_velocities(dt,am,vx,vy,vz,fx,fy,fz,fx_new,fy_new,fz_new):
    # at the point where a hop occured, subtract the included forces from an old state and replace them with forces of a new state
    for iat in range(0,len(am)):
        vx[iat] = vx[iat] + ( dt * (-fx[iat] + fx_new[iat]) / (2 * am[iat]) )
        vy[iat] = vy[iat] + ( dt * (-fy[iat] + fy_new[iat]) / (2 * am[iat]) )
        vz[iat] = vz[iat] + ( dt * (-fz[iat] + fz_new[iat]) / (2 * am[iat]) )
     #print(vz[iat])
    return(vx,vy,vz)   
    
def calc_forces(step, at_names, state, nstates, x, y, z, fx, fy, fz,
                pot_eners, ab_initio_file_path):
    """
    Call and collect an external script to calculate ab initio properties (force, energies)
    state = current state - PES for the forcess calc 
    """
    natoms = len(x)

    # Create geom file for which the forces will be calculated
    abinit_geom_file = "abinit_geom.xyz"
    with open (abinit_geom_file, "w") as agf:
         for iat in range(0,len(x)):
             line = ("".join("%2s %2.16e %2.16e %2.16e\n" % (at_names[iat],x[iat]/ANG_BOHR,y[iat]/ANG_BOHR,z[iat]/ANG_BOHR)))
             agf.write(line)
    agf.closed

    # CALL EXTERNAL HANDLING AB-INITIO CALCULATIONS       
    # electronic states starts from 1(ground state) 
    # while code starts from 0 index due to python 
    state = state + 1 
    abinit_inputs = "{} {}  {}  {} {} {}".format(ab_initio_file_path, abinit_geom_file, natoms, state, nstates, step)

    try:
        abinit_proc = subprocess.run(abinit_inputs, stdout=None, stderr=subprocess.PIPE, shell = True, check = True)	
    except subprocess.CalledProcessError as cpe: 
        print("Return code: {}\nError: {}".format(cpe.returncode, cpe.stderr))
        error_exit(4)

    # COLLECT DATA (TODO: diabatization - needs more forces)
    if  re.search(r'g09',ab_initio_file_path):
            grad = 1   # gaus exports forces -1
    elif re.search(r'molpro',ab_initio_file_path):
            grad = -1  #molpro export gradient
    with open ("gradients.dat", "r") as gef:
           
        for st in range(0, nstates):
            pot_eners[st] = float(gef.readline()) 
        for iat in range(0, natoms):
            line = gef.readline().split(" ")
            # FX FY FZ, gradient to forces 
            # TO DO Gaussian has forces, molpro gradients
            fx[iat] = grad * np.float64(line[0])
            fy[iat] = grad * np.float64(line[1])
            fz[iat] = grad * np.float64(line[2])    
    gef.closed
    return(fx , fy, fz, pot_eners)

def calc_energies(
    step, time, natoms, am, state, pot_eners, 
    vx, vy, vz, Etot_init, Etot_prev, ener_thresh):
     
    Ekin = 0.000
    for iat in range(0,natoms):
        vv = vx[iat] ** 2 + vy[iat] ** 2 + vz[iat] ** 2
        Ekin = Ekin + (0.5 * am[iat] * vv)
        Epot = pot_eners[state]  # state 0(GS), 1 (1.ex. state),..
        Etot = Ekin + Epot
        dE = (Etot - Etot_init)
        if (dE * AU_EV) >= ener_thresh: 
            print("Total energy change since start {} ".format(dE * AU_EV),
                  "larger then threshold {}.".format(ener_thresh))
            error_exit(7)
        dE_step = (Etot - Etot_prev)
    print_energies(step, time, Ekin, Epot, Etot, dE, dE_step)
    print_pes(time, step, pot_eners)

    return(Ekin, Epot, Etot, dE, dE_step)

def rescale_velocities(vx, vy, vz, v_scaling_fac):

    vx = [xx * v_scaling_fac for xx in vx] 
    vy = [yy * v_scaling_fac for yy in vy] 
    vz = [zz * v_scaling_fac for zz in vz] 

    return(vx, vy, vz)
    # Windows installed ubuntu has rather complicated path
    # if re.search(r'win',sys.platform):
    # testpath = "/mnt/c/Users/chalabaj/Documents/Coding/snafu-master/ABINITIO/test.sh" 
    # abinit_inputs = "wsl {} {}  {}  {}  {}".format(testpath, abinit_geom_file, natoms, state, nstates, step)
    # elif re.search(r'linux',sys.platform):  
