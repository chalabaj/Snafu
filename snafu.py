# SYSTEM IMPORTS:
import math
import sys
import os
import random
import time
import re
import numpy as np
from datetime import datetime

startTime = datetime.now()

# ENVIRONMENT LAYER & LOCAL IMPORT
# find local modules
# SNAFU_DIR=/path/to/SNAFU/snafu variable must be set before launch
# either export directly or in .bashrc SNAFU_DIR
# launcher enter to scratch dir with copied input files
cwd = os.getcwd()
sys.path.append(cwd)

modules_files = [
    'inits.py',
    'masses.py',
    'errors.py',
    'prints.py',
    'propagates.py',
    'landauzener.py',
    'constants.py',
    'restart.py'
]

# if some of the imports fail in modules ImportError rais None name
try:
    SNAFU_EXE = os.environ['SNAFU_DIR']
    sys.path.append(os.path.join(SNAFU_EXE, "snafu"))
    sys.path.append(SNAFU_EXE)
    from inits import (
        file_check, read_input, check_output_file, read_geoms, read_velocs,
        com_removal, init_fep_arrays
    )
    from masses import assign_masses
    from errors import error_exit
    from prints import (
        print_positions, print_velocities, print_snafu,
        print_state
    )
    from propagates import (
        calc_forces, calc_energies,
        update_velocities, update_positions, rescale_velocities,
        adjust_velocities
    )
    from landauzener import (
        calc_hopp
    )
    from restarts import (
        print_restart, check_restart_files, read_restart
    )
    from constants import *   #  import conversion factors and default values
    
    from defaults import *    # import all defualt values, only here otherwise could overwritte in some modules
    #from tera_propagates import (
    #    finish_tera, exit_tera, tera_connect, tera_init
    #)
except ImportError as ime:
    # module could have been removed or different module name, e.g. renamed in module file
    if ime.name is None:  
        print("Import in some modules {}".format(ime),
              "in snafu dir failed. Exiting...")
        exit(1)
    else:
        print("Module {} not found.".format(ime.name),
              "Make sure that {} contains snafu folder".format(SNAFU_EXE),
              "with: {}".format('\n'.join(modules_files)))
        exit(1)
except KeyError as ke:
    print("SNAFU_DIR is not set.",
          "See '.bashrc' file in your home directory",
          "or use 'env' command and make sure $SNAFU_DIR is exported.",
          "\nHint: export SNAFU_DIR=/path/to/SNAFU")
    exit(1)
else:
    print("All modules loaded succesfully.\n \n")
    print_snafu()

liner = ("_") * 100
# ---------------INIT--------------------------------------------------------

if __name__ == "__main__":

    print("Simulation started at: {}".format(startTime),
          "\nPython base: {}".format(sys.base_exec_prefix),
          "version: {}".format(sys.version[:5]),
          "\nSystem platform: {}".format(sys.platform),
          "\nRunning executable: {}".format(sys.path[-1])
          )
    # local runs dont create HOST env var, qsub SGE system does
    try:
        print("Working directory: {}".format(cwd),
              "on {}.".format(os.environ['HOSTNAME']))
    except KeyError:
        print("Working directory: {}".format(cwd))
    print(liner)

    # FILE CHECK - OBTAIN PATH TO FILES
    input_file_path, geom_file_path, vel_file_path, init_vel = file_check(cwd)
    print(liner)

    #  READ INPUT OPTIONS AND SET THEM AS VARIABLES:
    input_vars, ab_initio_file_path = read_input(cwd, input_file_path)
    #  Need to pass variables to function so that modules wont used defaults from constats mod
    globals().update(input_vars)
    try:
        natoms = int(natoms)
        maxsteps = int(maxsteps)
        state = int(init_state)   # initial or restart state
        dt = float(timestep)
        nstates = int(nstates)
        ener_thresh = float(ener_thresh)
        hop_thresh = float(hop_thresh)
        vel_adj = int(vel_adj)
        restart = int(restart)
        restart_freq = int(restart_freq)
        tera_mpi = int(tera_mpi)
        write_freq = int(write_freq)
    except ValueError as VE:
        print(VE)
        error_exit(9, " ")

    fx, fy, fz, fx_new, fy_new, fz_new, \
    pot_eners, x_new, y_new, z_new = init_fep_arrays(natoms, nstates)

    # READ INITIAL OR RESTART DATA
    rst_file_path = check_restart_files(restart, cwd)
    print(liner)
    if restart == 0:
        at_names, x, y, z  = read_geoms(natoms, geom_file_path)
        vx, vy, vz = read_velocs(init_vel, natoms, vel_file_path)

        # OBTAIN ATOMIC MASSES:
        masses = assign_masses(at_names)
        am = [mm * AMU for mm in masses]  # atomic mass units conversion
        # CENTER OF MASS REMOVAL 
        x, y, z = com_removal(x, y, z, am)
        # CALC INITIAL ENERGIES AND GRADIENTS
        fx, fy, fz, pot_eners = calc_forces(step, at_names, state, nstates,
                                            x, y, z, fx, fy, fz, pot_eners,
                                            ab_initio_file_path)
        pot_eners_array = np.copy(pot_eners)      
        
        Ekin, Epot, Etot, dE, dE_step = calc_energies(step, sim_time, natoms, am,
                                                      state, pot_eners,
                                                      vx, vy, vz, Etot_init,
                                                      Etot_prev, ener_thresh)
        Etot_init = Etot
        init_step = 1
    else:
        init_step, at_names, state, \
        x, y, z, vx, vy, vz, fx, fy, fz, \
        Ekin, Epot, Etot, Etot_init, \
        pot_eners_array = read_restart(rst_file_path, natoms)
        
        masses = assign_masses(at_names)
        am = [mm * AMU for mm in masses]  # atomic mass units conversion
        init_step = init_step + 1

    check_output_file(cwd, natoms, restart, init_step)
    
    print("Initial geometry:\n",
          "At    X         Y         Z         MASS:")
    xx = (x*BOHR_ANG).tolist()
    yy = (y*BOHR_ANG).tolist()
    zz = (z*BOHR_ANG).tolist()
    for iat in range(0, natoms):
        print("{} {:12.8f}".format(at_names[iat], xx[iat]),
              "{:12.8f} {:12.8f}".format(yy[iat], zz[iat]),
              "{:12.8f}".format(masses[iat]))

    print("Initial velocities:\n",
          "At    VX       VY       VZ         ")
    for iat in range(0, natoms):
        print("".join("%2s" " " "%3.3f" % (at_names[iat], vx[iat])),
              " %3.6f %2.6f " % (vy[iat], vz[iat]))
    print("{}".format(liner),
          "\nStep    Time/fs  dE_drift/eV   dE_step/eV    Hop  State") 
    
    #if tera_mpi:
    #    byte_coords = np.dstack((x,y,z))  # x,y,z must stack to single array
    #    comm = tera_connect()
    #    MO, MO_old, CiVecs, CiVecs_old, NAC, blob, \
    #    blob_old, SMatrix, civec_size, nbf_size,  \
    #    blob_size, TDip, Dip = tera_init(comm, at_names, natoms, nstates,
    #                                     byte_coords) 
    print(liner)
    with open('movie.xyz', 'a') as mov_file, \
         open('energies.dat', 'a') as eners_file, \
         open('PES.dat', 'a') as pes_file, \
         open('velocities.dat', 'a') as vel_file, \
         open('state.dat', 'a') as state_file, \
         open('restart.in', 'w') as rsf_file:

        #-------------------MAIN LOOP-----------------------------------------
        
        for step in range(init_step, maxsteps + 1):
    
            x_new, y_new, z_new = update_positions(dt, am, 
                                                   x, y, z,
                                                   x_new, y_new, z_new, 
                                                   vx, vy, vz, 
                                                   fx, fy, fz)
    
            fx_new, fy_new, fz_new, pot_eners = calc_forces(step, at_names, 
                                                            state, nstates, 
                                                            x_new, y_new, z_new,
                                                            fx_new, fy_new, fz_new,
                                                            pot_eners,
                                                            ab_initio_file_path)
    
            if not method == "bomd":
                if step >= 2:
                    hop, outstate, v_scal_fac, prob = calc_hopp(method, state,
                                                                pot_eners, 
                                                                pot_eners_array,
                                                                Ekin, dt,
                                                                hop_thresh)
    
                    if hop:
                        state = outstate
                        # use XYZ from prev. step to cacl F for a new state
                        
                        fx_new, fy_new, fz_new, pot_eners = calc_forces(
                            step, at_names, state, nstates, 
                            x, y, z,
                            fx_new, fy_new, fz_new, 
                            pot_eners, ab_initio_file_path)
    
                        #simple scaling or updatre velocities with new state forces
    
                        if not vel_adj:
                            vx, vy, vz = rescale_velocities(vx, vy, vz, v_scal_fac)
                        else:
                            vx, vy, vz = adjust_velocities(dt, am,
                                                           vx, vy, vz,
                                                           fx, fy, fz,
                                                           fx_new, fy_new, fz_new) 
    
                        # FXFYFZ for  XYZ(t = hop) in the new state already
                        fx = np.copy(fx_new)
                        fy = np.copy(fy_new)
                        fz = np.copy(fz_new)
                        EE = 0
                        for iat in range(0,natoms):
                            vvv = vx[iat] ** 2 + vy[iat] ** 2 + vz[iat] ** 2
                            EE = EE + (0.5 * am[iat] * vvv)
                        print("Old Ekin: {} \nScaled/Adjusted Ekin {}\n.".format(Ekin, EE))
    
                        # now finish the propagation step on new PES
    
                        x_new, y_new, z_new = update_positions(dt, am, 
                                                               x, y, z, 
                                                               x_new, y_new, z_new, 
                                                               vx, vy, vz, 
                                                               fx_new, fy_new,
                                                               fz_new)
    
    
                        fx_new, fy_new, fz_new, pot_eners = calc_forces(
                            step, at_names, state, nstates, 
                            x_new, y_new, z_new, fx_new, fy_new, fz_new,
                            pot_eners, ab_initio_file_path)
    
                    pot_eners_array = np.delete(pot_eners_array, 0, axis = 0)
                    pot_eners_array = np.vstack((pot_eners_array, pot_eners))  #  keep last two steps
    
                else:
                    pot_eners_array = np.vstack((pot_eners_array, pot_eners)) 
    
            vx, vy, vz = update_velocities(dt, am, 
                                           vx, vy, vz,
                                           fx, fy, fz,
                                           fx_new, fy_new, fz_new)
                
            fx = np.copy(fx_new)
            fy = np.copy(fy_new)
            fz = np.copy(fz_new)
            x = np.copy(x_new)
            y = np.copy(y_new)
            z = np.copy(z_new)
    
            sim_time = step * dt * AU_FS
            Etot_prev = Etot
            Ekin, Epot, Etot, dE, dE_step = calc_energies(step, sim_time, natoms, am,
                                                          state, pot_eners,
                                                          vx, vy, vz, Etot_init,
                                                          Etot_prev, ener_thresh)
    
            # print("Ekin {}, Epot {}, Etot{}".format(Ekin, Epot, Etot))
            print(" {:<6d}  {:<7.4f}  {:<12.4f}".format(step, sim_time, dE * AU_EV),
                  " {:<13.4f}".format(dE_step * AU_EV),
                  "{}     {}".format(str(hop)[0], state))
            #print("-----------------------------------------------------")
    
            # SAVE POSITION, VELOCITIES, ENERGIES AND RESTART
            if (step%write_freq == 0):
                print_energies(step, time, Ekin, Epot, Etot, dE, dE_step, eners_file)
                print_pes(time, step, pot_eners, pes_file)
                print_positions(step, sim_time, natoms, at_names, x, y, z, mov_file)
                print_velocities(step, sim_time, natoms, at_names, vx, vy, vz, vel_file)
                print_state(step, sim_time, state, state_file)
            
            print_restart(step, sim_time, natoms, at_names, state, timestep,
                          x, y, z, vx, vy, vz, fx, fy, fz,
                          Ekin, Epot, Etot, Etot_init, pot_eners_array, restart_freq, rsf_file)
    # FINAL PRINTS
    print(liner)
    print("#####JOB DONE.############")

    print("See output files:",
          "\nmovie.xyz, velocities.xyz,\nPES.dat, energies.dat,\nstate.dat")
    print(liner)
    stopTime = datetime.now()
    calc_time = (datetime.now() - startTime)
    print("Simulation ended at: {}".format(stopTime))
    print("Overall simulation sim_time (hh:mm:ss): {}".format(calc_time))
exit(0)
