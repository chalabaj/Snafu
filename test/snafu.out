Starting SNAFU.

Python exec:  /home/chalabaj/prog/Anaconda3-5.1.0-Linux-x86_64 ver: 3.6.5
System platform:  linux
Working directory:  /scratch/chalabaj/SNAFU-127054
_______________________________________________
No initial velocities, continue with zero initial velocities.
Simulation will start with parameters: 
natoms: 3
nstates: 1
init_state: 1
timestep: 4
maxsteps: 100
method: adiabatic
abinitio: test
restart: 0
_______________________________________________
atomic masses:
 [21894.7137735, 21894.7137735, 21894.7137735]
Molecular systems:
At  Mass     X     Y     Z:
 C 12.01  0.0000 0.0000 0.0000
 C 12.01  3.7795 0.0000 0.0000
 C 12.01  0.0000 3.7795 0.0000
_______________________________________________
Return code: 6
Error: b'grep: out: No such file or directory\n'
---------------------------------
Program was terminated due to an error! Exiting
4 - Ab initio code failed.
